src/traversals/iterators.jl

```diff
@@ -0,0 +1,130 @@
+"""
+    abstract type IteratorAlgorithm
+
+`IteratorAlgorithm` is an abstract type which specifies a particular algorithm to use when iterating through a graph.
+"""
+abstract type IteratorAlgorithm end
+
+
+"""
+    struct DFSIterator <: IteratorAlgorithm
+
+`DFSIterator` is a struct which specifies using depth-first traversal to iterate through a graph. A source node must be supplied to construct this iterator as `DFSIterator(g::AbstractGraph, source::Int)`.
+    
+# Examples
+```julia-repl
+julia> g = smallgraph(:house)
+{5, 6} undirected simple Int64 graph
+
+julia> for node in DFSIterator(g, 1)
+           display(node)
+       end
+1
+2
+4
+3
+5
+```
+"""
+struct DFSIterator <: IteratorAlgorithm
+    graph::AbstractGraph
+    source::Int
+end
+
+
+"""
+    struct BFSIterator <: IteratorAlgorithm
+
+`BFSIterator` is a struct which specifies using breadth-first traversal to iterate through a graph. A source node must be supplied to construct this iterator as `BFSIterator(g::AbstractGraph, source::Int)`.
+        
+# Examples
+```julia-repl
+julia> g = smallgraph(:house)
+{5, 6} undirected simple Int64 graph
+
+julia> for node in BFSIterator(g, 1)
+           display(node)
+       end
+1
+2
+3
+4
+5
+```
+"""
+struct BFSIterator <: IteratorAlgorithm
+    graph::AbstractGraph
+    source::Int
+end
+
+
+"""
+    abstract type AbstractGraphIteratorState
+
+`BasicGraphIteratorState` is an abstract type to hold the current state of iteration which is need for Julia's Base.iterate() function.
+"""
+abstract type AbstractGraphIteratorState end
```

Comment:
Is there any advantage of declaring an abstract type here, especially when we only have one subtype?

Reply:
Well, he applied my review. This leaves an API for Iterators that would store more information, for example current depth of the vertex, visited edges... I think there should be some work to define a clear API for all the types of iterators we would like to support.

Reply:
Yeah I think we need an API because I'd also like to implement some iterators for weighted graphs to do minimum spanning trees for instance. I don't know the algorithms for that, but I'd imagine that would require some more information to be stored in the state.